compiler/typecheck/TcDaml.hs

```diff
@@ -95,12 +102,18 @@ customDamlError ct
   | TyConApp con [TyConApp target [], TyConApp choice [], result] <- ctev_pred (ctEvidence ct)
   , check ["DA.Internal.Desugar", "DA.Internal.Template.Functions"] "HasExercise" con
   = Just $ TriedExercise { target = tyConName target, choice = tyConName choice, result }
-  | TyConApp con [TyConApp target [], LitTy (StrTyLit methodName), result] <- ctev_pred (ctEvidence ct)
+  | TyConApp con (_ `Snoc` TyConApp target [] `Snoc` LitTy (StrTyLit methodName) `Snoc` result) <- ctev_pred (ctEvidence ct)
```

Comment:
I'm probably missing something but I don't see what we gain from matching the end of the list. Did you find any cases where `HasMethod` had more than three arguments?

Reply:
Yeah, oddly enough the trace shows that some magic type primitives are being attached to the HasMethod instance's TyCon during type checking, and that was preventing matching. When HasMethod is pretty-printed, these are omitted, but an explicit print showed them

I want to delve deeper into this later, but for now this solves the issue